File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTrustAccount.java
package uk.gov.justice.hmpps.prison.repository.jpa.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.IdClass;
import javax.persistence.Table;
import java.io.Serializable;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@IdClass(OffenderTrustAccount.Pk.class)
@Table(name = "OFFENDER_TRUST_ACCOUNTS")
public class OffenderTrustAccount {

    @Id
    @Column(name = "CASELOAD_ID", nullable = false, insertable = false, updatable = false)
    private String prisonId;

    @Id
    @Column(name = "OFFENDER_ID", nullable = false, insertable = false, updatable = false)
    private Long offenderId;

    @Id
    @Column(name = "ACCOUNT_CLOSED_FLAG", nullable = false)
    private String accountClosedFlag;

    public boolean isAccountClosed() {
        return "Y".equals(accountClosedFlag);
    }

    @EqualsAndHashCode
    @NoArgsConstructor
    @AllArgsConstructor
    @ToString
    public static class Pk implements Serializable {
        private String prisonId;
        private Long offenderId;
    }
}

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderTrustAccount.kt
package uk.gov.justice.hmpps.prison.repository.jpa.model

import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Embeddable
import jakarta.persistence.EmbeddedId
import jakarta.persistence.Entity
import jakarta.persistence.Table
import org.hibernate.Hibernate
import org.hibernate.type.YesNoConverter
import uk.gov.justice.hmpps.prison.repository.jpa.helper.EntityOpen
import java.io.Serializable
import java.math.BigDecimal
import kotlin.jvm.javaClass

@Embeddable
class OffenderTrustAccountId(
  @Column(name = "CASELOAD_ID", nullable = false)
  val prisonId: String,

  @Column(name = "OFFENDER_ID", nullable = false)
  val offenderId: Long,
) : Serializable

@Entity
@EntityOpen
@Table(name = "OFFENDER_TRUST_ACCOUNTS")
data class OffenderTrustAccount(

  @EmbeddedId
  val id: OffenderTrustAccountId,

  @Column(name = "ACCOUNT_CLOSED_FLAG", nullable = false)
  @Convert(converter = YesNoConverter::class)
  val accountClosed: Boolean,

) {

  @Column(name = "HOLD_BALANCE")
  var holdBalance: BigDecimal? = null

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
    other as OffenderTrustAccount
    return id == other.id
  }

  override fun hashCode(): Int = javaClass.hashCode()
}
